pkg/ingester/errors_test.go

```diff
@@ -348,151 +348,305 @@ func TestWrapOrAnnotateWithUser(t *testing.T) {
+			expectedMessage: newPerMetricMetadataLimitReachedError(10, labels).Error(),
+			expectedDetails: &mimirpb.WriteErrorDetails{Cause: mimirpb.BAD_DATA},
+		},
+		"a wrapped perMetricMetadataLimitReachedError gets translated into a non-loggable errorWithStatus FailedPrecondition error with details": {
+			err:             fmt.Errorf("wrapped: %w", newPerMetricMetadataLimitReachedError(10, labels)),
+			expectedCode:    codes.FailedPrecondition,
+			expectedMessage: fmt.Sprintf("wrapped: %s", newPerMetricMetadataLimitReachedError(10, labels).Error()),
+			expectedDetails: &mimirpb.WriteErrorDetails{Cause: mimirpb.BAD_DATA},
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			handledErr := handlePushErrorWithGRPC(tc.err)
+			stat, ok := status.FromError(handledErr)
+			require.True(t, ok)
+			require.Equal(t, tc.expectedCode, stat.Code())
+			require.Equal(t, tc.expectedMessage, stat.Message())
+			checkErrorWithStatusDetails(t, stat.Details(), tc.expectedDetails)
+			if tc.doNotLogExpected {
+				var doNotLogError log.DoNotLogError
+				require.ErrorAs(t, handledErr, &doNotLogError)
+				require.False(t, doNotLogError.ShouldLog(context.Background(), 0))
+			}
+		})
+	}
+}
+
+func TestHandlePushErrorWithHTTPGRPC(t *testing.T) {
+	originalMsg := "this is an error"
+	originalErr := errors.New(originalMsg)
+	labelAdapters := []mimirpb.LabelAdapter{{Name: labels.MetricName, Value: "testmetric"}, {Name: "foo", Value: "biz"}}
+	labels := mimirpb.FromLabelAdaptersToLabels(labelAdapters)
 	timestamp := model.Time(1)
-	testCases := []struct {
-		name                string
+
+	testCases := map[string]struct {
 		err                 error
 		doNotLogExpected    bool
 		expectedTranslation error
 	}{
-		{
-			name:                "a generic error is not translated",
+		"a generic error is not translated": {
 			err:                 originalErr,
 			expectedTranslation: originalErr,
 		},
-		{
-			name:                "a DoNotLog error of a generic error is not translated",
+		"a DoNotLog error of a generic error is not translated": {
 			err:                 log.DoNotLogError{Err: originalErr},
 			expectedTranslation: log.DoNotLogError{Err: originalErr},
 			doNotLogExpected:    true,
 		},
-		{
-			name:                "an unavailableError gets translated into an errorWithStatus Unavailable error",
+		"an unavailableError gets translated into an errorWithStatus Unavailable error": {
 			err:                 newUnavailableError(services.Stopping),
 			expectedTranslation: newErrorWithStatus(newUnavailableError(services.Stopping), codes.Unavailable),
 		},
-		{
-			name: "a wrapped unavailableError gets translated into a non-loggable errorWithStatus Unavailable error",
-			err:  fmt.Errorf("wrapped: %w", newUnavailableError(services.Stopping)),
+		"a wrapped unavailableError gets translated into a non-loggable errorWithStatus Unavailable error": {
+			err: fmt.Errorf("wrapped: %w", newUnavailableError(services.Stopping)),
```

Comment:
Nice, really like these comprehensive test cases.

Reply:
Thank you :)